### DISCUSSION TITLE
Discussion point: play-haskell as an examination tool

### DISCUSSION THREAD

**Discussion by sandsify** [haskell/play-haskell#44] [General]
Not really an issue: I'm interested in the possibilities of using play-haskell as an examination tool as part of electronic (sandboxed) course examination, and generally as a teaching tool. 

Particularly since this is the first system that seems to support imports of standard modules out-of-the-box (great stuff!) comared to all the other systems that I looked at previously.

Perhaps there is a forum relevant for such discussions?

**Comment by tomsmeding** [comment]
Fair point about the forum -- I enabled github discussions on the repo and transferred this here. @sandsify It does seem more appropriate than an issue.

I have a private project that reuses the playground backend for an assignment submission tool. Its frontend is changed from the interactive playground interface to just a submission page, though, and it needs some maintenance/generalisation work (i.e. less hard-coding) to make it suitable for general consumption.

What exactly were you thinking of?

**Reply by sandsify** [reply]
Well, I can tell you about the use-case that I am thinking of.  We use an online exam system (inspera).  Students bring laptops with a special browser that locks their machine during the exam.   But you have the possibility to whitelist some web pages which can then be accessed during the exam.  In the exam setup I could whitelist play.haskell.org (only) and they could develop code then paste their answers into the exam when they are done.  

But going beyond that, it might be nice if they saved their solutions and also pasted the url into the submission.  Then we could run their solutions with two clicks.  But that maybe raises a bunch of problems - can they now update the submission after the exam (maybe not)?  But in any case in the current system we would have to whitelist everything under https://play.haskell.org/saved/ - and this would open up for them access code saved in advance of the exam (e.g. solutions to old exam questions, notes).    What we need is some kind of write-only saves (allowing student to generate a save link without actually visiting the saved subdomain)

For educational purposes it would be nice to support some semi-automated testing by creation of a test-stub that we can easily combine with a saved solution, but I have not thought much about that yet.  

Anyway - thanks for creating a discussion area.

**Reply by tomsmeding** [reply]
How difficult these things would be depends mostly on how generalised you want the functionality to be, I think.

It would for example be relatively easy to fork the playground and:
- put authentication (http basic auth is sufficient if the thing is served over https) on the save links, effectively making them write-only for the students
- ensure that the submissions can indeed not be changed after submission (saves are immutable, currently, and I don't plan on changing that, but with a fork you don't run the risk of future me changing my mind)
- add a 4th run variant (current variants are `run`, `core` and `asm`) that does whatever custom thing you like; through some indirections, the eventual code that's executed in the sandboxes is [this](https://github.com/haskell/play-haskell/blob/01b8bef952e6d0031c5216bb6ce3cf7bba653e63/play-haskell-worker/bwrap-files/stage-3.sh#L29-L54), so adding a 4th option that links the code against some other files wouldn't be hard. Legitimately the hardest part would be to ensure that the students' code can be run in the playground (i.e. it's a `Main` module) but also when linked against some testing framework (i.e. it's not `Main`). :P Perhaps you'd want to modify `run` as well to do whatever you want to the module the students write. Or maybe you want the testing framework to not be inside haskell but instead some shell script etc. that tests the student's program as a black box.
- keep the changes of the fork localised enough that merging in future development to the main playground is not too hard.

You probably also want to host your own version so that students' submissions don't end up on someone else's server, but on your own. Good for privacy regulations.

Or maybe this was not what you had in mind!
